main.go

```diff
@@ -52,82 +44,123 @@ func main() {
+	close(proxyChan)
 }
 
 // pass a stopChan to the go routtine
-func proxy(local io.ReadWriteCloser, remoteAddr *net.TCPAddr, stopChan chan struct{}) {
-	fmt.Printf("Opening a new connection on remoteAddr, %s\n", remoteAddr)
-	remote, err := net.DialTCP("tcp", nil, remoteAddr)
+func proxy(client *net.TCPConn, redisAddr *net.TCPAddr, stopChan <-chan string) {
+	redis, err := net.DialTimeout("tcp4", redisAddr.String(), 50*time.Millisecond)
 	if err != nil {
-		log.Println(err)
-		local.Close()
+		log.Printf("[PROXY %s => %s] Can't establish connection: %s\n", client.RemoteAddr().String(), redisAddr.String(), err)
+		client.Close()
 		return
 	}
-	go pipe(local, remote)
-	go pipe(remote, local)
-	<-stopChan // read from stopChan
-	fmt.Println("Closing Proxy")
-	local.Close()
-}
 
-func getMasterAddr(sentinelAddress *net.TCPAddr, masterName string) (*net.TCPAddr, error) {
-	conn, err := net.DialTCP("tcp", nil, sentinelAddress)
-	if err != nil {
-		return nil, err
-	}
+	log.Printf("[PROXY %s => %s] New connection\n", client.RemoteAddr().String(), redisAddr.String())
+	defer client.Close()
+	defer redis.Close()
 
-	defer conn.Close()
+	clientChan := make(chan string)
+	redisChan := make(chan string)
 
-	conn.Write([]byte(fmt.Sprintf("sentinel get-master-addr-by-name %s\n", masterName)))
+	go pipe(client, redis, redisChan)
+	go pipe(redis, client, clientChan)
 
-	b := make([]byte, 256)
-	_, err = conn.Read(b)
-	if err != nil {
-		log.Fatal(err)
+	select {
+	case <-stopChan:
+	case <-clientChan:
+	case <-redisChan:
 	}
 
-	parts := strings.Split(string(b), "\r\n")
+	log.Printf("[PROXY %s => %s] Closing connection\n", client.RemoteAddr().String(), redisAddr.String())
+}
 
-	if len(parts) < 5 {
-		err = errors.New("Couldn't get master address from sentinel")
-		return nil, err
-	}
+func getMasterAddr(sentinelAddress string, masterName string) (*net.TCPAddr, error) {
 
-	//getting the string address for the master node
-	stringaddr := fmt.Sprintf("%s:%s", parts[2], parts[4])
-	addr, err := net.ResolveTCPAddr("tcp", stringaddr)
+	sentinelHost, sentinelPort, err := net.SplitHostPort(sentinelAddress)
+	if err != nil {
+		return nil, fmt.Errorf("Can't find Sentinel: %s", err)
+	}
 
+	sentinels, err := net.LookupIP(sentinelHost)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Can't lookup Sentinel: %s", err)
 	}
 
-	//check that there's actually someone listening on that address
-	conn2, err := net.DialTCP("tcp", nil, addr)
-	if err == nil {
+	for _, sentinelIP := range sentinels {
+		conn, err := net.DialTimeout("tcp4", sentinelIP.String()+":"+sentinelPort, 100*time.Millisecond)
+		if err != nil {
+			log.Printf("[MASTER] Unable to connect to Sentinel at %v:%v: %v", sentinelIP, sentinelPort, err)
+			continue
+		}
+		defer conn.Close()
+
+		conn.Write([]byte(fmt.Sprintf("sentinel get-master-addr-by-name %s\n", masterName)))
+
+		b := make([]byte, 256)
+		_, err = conn.Read(b)
+		if err != nil {
+			log.Fatal(err)
```

Comment:
Seems we've customized every other error in the implementation, any reason not to do the same here?